source/common/network/connection_impl.cc

```diff
@@ -115,6 +115,11 @@ ConnectionImpl::~ConnectionImpl() {
   // deletion). Hence the assert above. However, call close() here just to be completely sure that
   // the fd is closed and make it more likely that we crash from a bad close callback.
   close(ConnectionCloseType::NoFlush);
+
+  // Filters may care about whether this connection is an SSL connection or not in their
+  // destructors for stat reasons. We destroy the filters here vs. the base class destructors
+  // to make sure they have the chance to still inspect SSL specific data via virtual functions.
+  filter_manager_.destroyFilters();
```

Comment:
This is kind of confusing. We are in the base class, right? Is this needed anymore?

Reply:
I would need to look into this more, but I would triple check some cases where filters look at SSL status in their destructors to dec stats (HTTP connection manager does this for sure). Assuming that works, I think with the refactor this can likely be deleted outright.

Reply:
Let me double check, I added this to base class since I saw some test failure.

Reply:
I think re-order the members of Connection impl (i.e. TransportSocketPtr before FilterManager) would resolve this, will run test and update the PR.

Reply:
Yup, +1, thanks